source/shared/core_stmt.cpp

```diff
@@ -2682,6 +2660,134 @@ bool reset_ae_stream_cursor( _Inout_ sqlsrv_stmt* stmt ) {
     return false;
 }
 
+void adjustInputPrecision( _Inout_ zval* param_z, _In_ SQLSMALLINT decimal_digits ) {
+    size_t maxDecimalPrecision = 38;
+    // maxDecimalStrLen is the maximum length of a stringified decimal number
+    // 6 is derived from: 1 for '.'; 1 for sign of the number; 1 for 'e' or 'E' (scientific notation);
+    //                    1 for sign of scientific exponent; 2 for length of scientific exponent
+    // if the length is greater than maxDecimalStrLen, do not change the string
+    size_t maxDecimalStrLen = maxDecimalPrecision + 6;
+    if (Z_STRLEN_P(param_z) > maxDecimalStrLen) {
+        return;
+    }
+    std::vector<size_t> digits;
+    char* ptr = ZSTR_VAL( Z_STR_P( param_z ));
+    bool isNeg = false;
+    char scientificChar = ' ';
+    short scientificExp = 0;
+    // parse digits in param_z into the vector digits
+    if( *ptr == '+' || *ptr == '-' ){
+        if( *ptr == '-' ){
+            isNeg = true;
+        }
+        ptr++;
+    }
+    short numInt = 0;
+    short numDec = 0;
+    while( isdigit( *ptr )){
+        digits.push_back( *ptr - '0' );
+        ptr++;
+        numInt++;
+    }
+    if( *ptr == '.' ){
+        ptr++;
+        while( isdigit(*ptr) ){
+            digits.push_back( *ptr - '0' );
+            ptr++;
+            numDec++;
+        }
+    }
+    if( *ptr == 'e' || *ptr == 'E' ){
+        scientificChar = *ptr;
+    }
+    if( scientificChar != ' ' ){
+        ptr++;
+        bool isNegExp = false;
+        if( *ptr == '+' || *ptr == '-' ){
+            if( *ptr == '-' ){
+                isNegExp = true;
+            }
+            ptr++;
+        }
+        while( isdigit( *ptr )){
+            scientificExp = scientificExp * 10 + ( *ptr - '0' );
+            ptr++;
+        }
+        SQLSRV_ASSERT( scientificExp <= maxDecimalPrecision, "Input decimal overflow: sql decimal type only supports up to a precision of 38." );
+        if( isNegExp ){
+            scientificExp = scientificExp * -1;
+        }
+    }
+    // if ptr is not pointing to a null terminator at this point, that means the decimal string input is invalid
+    // do not change the string and let SQL Server handle the invalid decimal string
+    // if number of decimal is less than the exponent, that means the number is a whole number, so no need to adjust the precision
+    if(*ptr == '\0' && ( int )numDec > scientificExp ){
+        int decToRemove = numDec - scientificExp - decimal_digits;
+        if( decToRemove > 0 ){
+            bool carryOver = false;
+            short backInd = 0;
+            // pop digits from the vector until there is only 1 more decimal place than required decimal_digits
+            while( decToRemove != 1 && !digits.empty() ){
+                digits.pop_back();
+                decToRemove--;
```

Comment:
I still think we should not continue parsing the input string when some digits are meant to be dropped (popped) later.

Reply:
let me look up the performance of finding a character ('e') in a string and get back to you.